### PR TITLE
[Feature] Application Role Connections Metadata support

### PR DESCRIPTION
**Pull request by Misha-133**
This PR adds support for the [Application Role Connections Metadata](https://discord.com/developers/docs/resources/application-role-connection-metadata)

### Changes
- [x] add data models
- [x] implement api methods

### CODE REVIEW

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 1]
```diff
@@ -2476,5 +2477,29 @@ private static string WebhookQuery(bool wait = false, ulong? threadId = null)
         }
 
         #endregion
+
+        #region Application Role Connections Metadata
+
+        public async Task<IEnumerable<RoleConnectionMetadata>> GetApplicationRoleConnectionMetadataRecordsAsync(RequestOptions options = null)
```
Comment (csmir) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: Missing xml comments

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 2]
```diff
@@ -2476,5 +2477,29 @@ private static string WebhookQuery(bool wait = false, ulong? threadId = null)
...
+        {
+            return await SendAsync<IEnumerable<RoleConnectionMetadata>>("GET", $"/applications/{CurrentApplicationId}/role-connections/metadata", options: options).ConfigureAwait(false);
+        }
+
+        public async Task<IEnumerable<RoleConnectionMetadata>> UpdateApplicationRoleConnectionMetadataRecordsAsync(IEnumerable<RoleConnectionMetadata> roleConnections, RequestOptions options = null)
```
Comment (csmir) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: Missing xml comments

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 3]
```diff
@@ -2476,5 +2477,29 @@ private static string WebhookQuery(bool wait = false, ulong? threadId = null)
...
+        {
+            return await SendJsonAsync<IEnumerable<RoleConnectionMetadata>>("PUT", $"/applications/{CurrentApplicationId}/role-connections/metadata", roleConnections, options: options).ConfigureAwait(false);
+        }
+
+        public async Task<RoleConnection> GetUserApplicationRoleConnection(RequestOptions options = null)
```
Comment (csmir) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: Missing xml comments

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 4]
```diff
@@ -2476,5 +2477,29 @@ private static string WebhookQuery(bool wait = false, ulong? threadId = null)
...
+        {
+            return await SendAsync<RoleConnection>("GET", $"/users/@me/applications/{CurrentApplicationId}/role-connection", options: options);
+        }
+
+        public async Task<RoleConnection> GetUserApplicationRoleConnection(RoleConnection connection, RequestOptions options = null)
```
Comment (csmir) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: Missing xml comments

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnectionMetadata.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Discord;
+
+public class RoleConnectionMetadata
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnectionMetadata.cs:9]: Missing xml comments

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+using System.Collections.Generic;
+
+namespace Discord;
+
+public class RoleConnection
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs:9]: Missing xml comments

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    /// <summary>
+    ///     Gets the username on the platform a bot has connected.
+    /// </summary>
+    public string PlatformUsername { get; }
+
+    /// <summary>
+    ///     
+    /// </summary>
+    public IReadOnlyDictionary<string, object> Metadata { get; }
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs]: Missing xml comments

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnectionMetadata.cs (hunk 5 above)
Comment (csmir) [reply src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnectionMetadata.cs:9]: Includes all below

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnectionMetadataType.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
+using System;
+
+namespace Discord;
+
+/// <summary>
+///     Represents the type of Application Role Connection Metadata
+/// </summary>
+public enum RoleConnectionMetadataType
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnectionMetadataType.cs:8]: Consider adding a dot at the end of all summaries for consistency

**Review by csmir — CHANGES_REQUESTED** [review]

**Comment by csmir** [comment]
Oh oops very sorry looked over the fact this is a draft, ignore review in continuation

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Discord;
+
+/// <summary>
+///     Represents the connection object that the user has attached.
+/// </summary>
+public class RoleConnection
+{
+    /// <summary>
+    ///     Gets the vanity name of the platform a bot has connected.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs]:
```suggestion
    ///     Gets the vanity name of the platform a bot has connected to.
```

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    public string PlatformName { get; }
+
+    /// <summary>
+    ///     Gets the username on the platform a bot has connected.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs]:
```suggestion
    ///     Gets the username on the platform a bot has connected to.
```

**Review by csmir — APPROVED** [review]

File: src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs (hunk 10 above)
Comment (Misha-133) [reply src/Discord.Net.Core/Entities/ApplicationRoleConnection/RoleConnection.cs]: not sure if this needs to be changed
this comment comes from [discord docs](
https://discord.com/developers/docs/resources/user#update-user-application-role-connection-json-params)

**Review by MrCakeSlayer — APPROVED** [review]

**Pull request merged**